### DISCUSSION TITLE
attiny1614 TCB0 W0 and TCB1 W0 not connecting to Pins

### DISCUSSION THREAD

**Discussion by farfarfromhome** [SpenceKonde/megaTinyCore#524] [Q&A] [answered]
This is my first time using the attiny1614 and I am stumped.  I am trying to output from Timer Counter B to a pin.  The PORT Function Multiplexing in the spec shows TCB1 WO connecting to PA3 and TCB0 WO to PA5.   I followed the guidelines in section 21.3.2 and set up a count of 16 and verified that the counter is working by reading from the registers and sampling the counter.  Programmed IO cannot keep up with the counter but all readings are in range and changing so it sure looks like it is working. 
`
Checking Counter TCB0
CLKCTRL_MCLKCTRLB 0x0
TCB0_CTRLA 0x3
TCB0_CTRLB 0x10
TCB0_CCMPL 0x10
TCB0_STATUS 0x1
TCB0_CNT : 0x10: 0x9: 0x1: 0xb: 0x3: 0xd: 0x5: 0xf: 0x7: 0x0: 0x9: 0x2: 0xb: 0x4: 0xd: 0x6
Checking Counter TCB1
TCB1_CTRLA 0x3
TCB1_CTRLB 0x10
TCB1_CCMPL 0x10
TCB1_STATUS 0x1
TCB1_CNT : 0xb: 0x4: 0xd: 0x6: 0xf: 0x8: 0x0: 0xa: 0x2: 0xc: 0x4: 0xe: 0x6: 0x10: 0x8: 0x1

`
I then checked the Port Registers and everything looks OK including the enabling of PA3 and PA5 as outputs! 
`
Checking Port Mux registers
PORTMUX_CTRLA 0x0
PORTMUX_CTRLB 0x1
PORTMUX_CTRLC 0x0
PORTMUX_CTRLD 0x0

Checking Port Registers
PORTA_DIR 0x2a
PORTA_OUT 0x4
PORTA_IN 0xd7
`
I checked my scope by setting up a PWM output on another pin and it works!  So I don't believe I have a clash with a library and megaTinyCore is doing everything correctly at the register level.  The attiny1614 errata does not indicate this is a problem.  I am hoping the experts here can point to what I am doing wrong!

**Comment by hmeijdam** [comment]
May I suggest you post your code here (within code tags), so others can try your code and test it.

**Reply by farfarfromhome** [reply]
Thanks for the quick reply.  Attached please find the test sketch.  There are three defines;-
1) REDIRECT_SERIAL selects alternate serial port mapping which I need for my board.
2) DUMP_PORTA_REGS prints the port control registers.
3) DUMP_TCB_REGS prints timerB control registers.
I turned of serial port re-mapping in the attached code.
`
#include <Arduino.h>

//#define REDIRECT_SERIAL         //Redirects Serial to alternate pins
#define DUMP_PORTA_REGS        //Display PORTA setup registers
#define DUMP_TCB_REGS          //Display TCBn registers

#define TCB0_WO PIN_PA5   //Pin 3 - TCB0
#define TCB1_WO PIN_PA3   //Pin 13 - TCB1

void setup() {
    #ifdef REDIRECT_SERIAL
        Serial.swap(1);
    #endif

    Serial.begin(115200);
    while (!Serial) { ; }
    Serial.println("TimerB Test");

    pinMode(TCB0_WO, OUTPUT);  // Initialize pin to output
    pinMode(TCB1_WO, OUTPUT);  // Initialize pin to output

    CLKCTRL_MCLKCTRLB  = 5;   // 0 - CLK_PER = CLK_MAIN, 5 - CLK_PER = CLK_MAIN/2

    // Counter TCB0 setup
    TCB0_CCMPL = 0x10;        //Top Count
    TCB0_CCMPH = 0;
    TCB0_CTRLB = 0x10;        //CMP0EN
    TCB0_CTRLA = 3;           // 1 - Enable Count & fTCA = fCLK_PER, 3 - Enable Count & fTCA = fCLK_PER/2

    // Counter TCB1 setup
    TCB1_CCMPL = 0x10;        //Top Count
    TCB1_CCMPH = 0;
    TCB1_CTRLB = 0x10;        //CMP0EN
    TCB1_CTRLA = 3;           // 1 - Enable Count & fTCA = fCLK_PER, 3 - Enable Count & fTCA = fCLK_PER/2
 }

void loop() {
    
    #ifdef DUMP_PORTA_REGS
        Serial.println("Checking Port Mux registers");   
        register16_t portmux_ctrla = PORTMUX_CTRLA;   
        register16_t portmux_ctrlb = PORTMUX_CTRLB;   
        register16_t portmux_ctrlc = PORTMUX_CTRLC;   
        register16_t portmux_ctrld = PORTMUX_CTRLD;
        Serial.printf("PORTMUX_CTRLA 0x%x\n\r", portmux_ctrla);
        Serial.printf("PORTMUX_CTRLB 0x%x\n\r", portmux_ctrlb);
        Serial.printf("PORTMUX_CTRLC 0x%x\n\r", portmux_ctrlc);
        Serial.printf("PORTMUX_CTRLD 0x%x\n\r", portmux_ctrld);

        Serial.println(""); 

        Serial.println("Checking Port Registers");   
        register16_t porta_dir = PORTA_DIR;
        register16_t porta_out = PORTA_OUT;
        register16_t porta_in = PORTA_IN;
        register16_t porta_intflags = PORTA_INTFLAGS;

        register16_t porta_pin0ctrl = PORTA_PIN0CTRL;
        register16_t porta_pin1ctrl = PORTA_PIN1CTRL;
        register16_t porta_pin2ctrl = PORTA_PIN2CTRL;
        register16_t porta_pin3ctrl = PORTA_PIN3CTRL;
        register16_t porta_pin4ctrl = PORTA_PIN4CTRL;
        register16_t porta_pin5ctrl = PORTA_PIN5CTRL;
        register16_t porta_pin6ctrl = PORTA_PIN6CTRL;
        register16_t porta_pin7ctrl = PORTA_PIN7CTRL;

        Serial.printf("PORTA_DIR 0x%x\n\r", porta_dir);
        Serial.printf("PORTA_OUT 0x%x\n\r", porta_out);
        Serial.printf("PORTA_IN 0x%x\n\r", porta_in);
        Serial.printf("PORTA_INFLAGS 0x%x\n\r", porta_intflags);

        Serial.printf("PORTA_PIN0CTRL 0x%x\n\r", porta_pin0ctrl);
        Serial.printf("PORTA_PIN1CTRL 0x%x\n\r", porta_pin2ctrl);
        Serial.printf("PORTA_PIN2CTRL 0x%x\n\r", porta_pin2ctrl);
        Serial.printf("PORTA_PIN3CTRL 0x%x\n\r", porta_pin3ctrl);
        Serial.printf("PORTA_PIN4CTRL 0x%x\n\r", porta_pin4ctrl);
        Serial.printf("PORTA_PIN5CTRL 0x%x\n\r", porta_pin5ctrl);
        Serial.printf("PORTA_PIN6CTRL 0x%x\n\r", porta_pin6ctrl);
        Serial.printf("PORTA_PIN7CTRL 0x%x\n\r", porta_pin7ctrl);

        Serial.println("");  
    #endif 
    
    #ifdef DUMP_TCB_REGS
        Serial.println("Checking Counter TCB0");
        register8_t mclkctrlb = CLKCTRL_MCLKCTRLB;
        register8_t tcb0_ctrla = TCB0_CTRLA;
        register8_t tcb0_ctrlb = TCB0_CTRLB;
        register8_t tcb0_ccmpl = TCB0_CCMPL;  
        register8_t tcb0_status = TCB0_STATUS;
        register16_t tcb0_countl[32] = {0};
        for (int i = 0; i < 16; i++)
            tcb0_countl[i] = TCB0_CNT;

        Serial.printf("CLKCTRL_MCLKCTRLB 0x%x\n\r", mclkctrlb);
        Serial.printf("TCB0_CTRLA 0x%x\n\r", tcb0_ctrla); 
        Serial.printf("TCB0_CTRLB 0x%x\n\r", tcb0_ctrlb);
        Serial.printf("TCB0_CCMPL 0x%x\n\r", tcb0_ccmpl);  
        Serial.printf("TCB0_STATUS 0x%x\n\r", tcb0_status);
        Serial.print("TCB0_CNT ");
        for (int i = 0; i < 16; i++)
            Serial.printf(": 0x%x", tcb0_countl[i]);
        Serial.println("");

        Serial.println("Checking Counter TCB1");
        register8_t tcb1_ctrla = TCB1_CTRLA;
        register8_t tcb1_ctrlb = TCB1_CTRLB;
        register8_t tcb1_ccmpl = TCB1_CCMPL; 
        register8_t tcb1_status = TCB1_STATUS;
        register16_t tcb1_countl[32] = {0};
        for (int i = 0; i < 16; i++)
            tcb1_countl[i] = TCB1_CNT;

        Serial.printf("TCB1_CTRLA 0x%x\n\r", tcb1_ctrla); 
        Serial.printf("TCB1_CTRLB 0x%x\n\r", tcb1_ctrlb);
        Serial.printf("TCB1_CCMPL 0x%x\n\r", tcb1_ccmpl);
        Serial.printf("TCB1_STATUS 0x%x\n\r", tcb1_status);
        Serial.print("TCB1_CNT ");
        for (int i = 0; i < 16; i++)
            Serial.printf(": 0x%x", tcb1_countl[i]);
        Serial.println("");
    #endif

    delay(10000);
}
`

**Comment by SpenceKonde** [comment]
The behavior you describe is expected with those settings. You have not instructed the timers to output PWM. 

1. CTRLB is set to 0x10, meaning that CNTMODE = 0 = periodic interrupt mode. This mode does not use the compare output pin. Only modes 6 (single shot mode) and 7 (8-bit pwm mode) use the output pin. 
2. You are writing the CCMPH and CCMPL registers separately. Per the silicon errata datasheet, you must write to them as a single 16-bit register, even when they are otherwise independent registers (when CNTMODE = 7 (8-bit PWM mode). The same applies to reading these registers.
3. Even if you were in PWM mode, the values you supply for CCMPH and CCMPL do not make sense for generating PWM. 

The type B timers on modernAVRs REALLY SUCK for generating PWM, and even if they worked as described in the datasheet, rather than the errata, they would still suck. They are not made for that. The type A and type D timers are the ones the designers intended to be powerful PWM generation tools.

**Reply by SimonMerrett** [reply]
Spence, I have to ask, how do you know this? I just wasted several hours trying to get  a waveform (default periodic interrupt mode) on TCB0 WO. When I should have searched the discussions and issues in this repo far sooner. You have to admit, the datasheet does give the impression that WO would be toggling away in periodic interrupt mode:
```
21.3.2 Initialization
By default, the TCB is in Periodic Interrupt mode. Follow these steps to start using it:
   1. Write a TOP value to the Compare/Capture (TCBn.CCMP) register.
   2. Optional: Write the Compare/Capture Output Enable (CCMPEN) bit in the Control B (TCBn.CTRLB) register
   to‘1’. This will make the waveform output available on the corresponding pin, overriding the value in the
   corresponding PORT output register.
   3. Enable the counter by writing a ‘1’ to the ENABLE bit in the Control A (TCBn.CTRLA) register.
   The counter will start counting clock ticks according to the prescaler setting in the Clock Select (CLKSEL) bit
   field in the Control A (TCBn.CTRLA) register.
   4. The counter value can be read from the Count (TCBn.CNT) register. The peripheral will generate a CAPT
   interrupt and event when the CNT value reaches TOP.
      4.1. If the Compare/Capture register is modified to a value lower than the current Count register, the
      peripheral will count to MAX and wrap around.
```
But alas, this does nothing and also prevents other code (direct port manipulation on those pins - which perhaps I shouldn't be doing but I wanted any output for validation) from working. I call it misleading datasheet writing. Thank goodness for this repo.

**Comment by farfarfromhome** [comment]
I have found an example on Hackaday that uses the CCL to output a combination of TimerB and TimerD to PA7 that works!  Maybe it will give me a clue to what is wrong or another way to achieve my goal!

**Comment by farfarfromhome** [comment]
Thanks for the confirmation.  I tried the interrupt mode with the CCL and the same thing happens, the output is not passed through to WO.  I guess this makes sense since otherwise output pins would be wastefully consumed.  Actually I just want a very fast output clock to drive power line communication.  I will take a look at timerD but actually the CCL maybe the way to go.  If not its a board mod to add an external xtal, yuck.  The other end of the communication uses an ESP32 and it was three lines of code and five minutes to achieve!

I really appreciate you spending time to respond, thank you.

**Comment by SpenceKonde** [comment]
Why not use the TCB in PWM8 mode, then?
as long as the required frequency is less than 512x the system clock, PWM8 mode should be fine...

**Comment by farfarfromhome** [comment, marked as answer]
Unfortunately that isn't fast enough, I need between 6 and 9 MHZ.   Hey, I have CCL working now, using J-K flip flop (divides by 2) clocked off the internal clock.  Using 16MHz will give me 8MHz so job done.  Only problem is the wrong pin so tiny wires for now and PCB update in any case.

**Reply by SpenceKonde** [reply]
You can get 8 MHz from a TCB in PWM8 mode too? 

clock it from main clock instead of main clock divided by 2, period of 1 (so 2 clocks, -> 8 MHz) duty cycle 0 (so 1 clock)... am I missing something?

**Reply by farfarfromhome** [reply]
I didn't try that ...... after realizing the output in interrupt mode didn't go to the pin I then looked at the CCL.  It seemed like an elegant solution with no timers and potential library interaction to worry about.  I will try the PWM8 for the sake of my pin and report back.

**Reply by SpenceKonde** [reply]
just remember that you need to set CCMP as a 16-bit value. not a pair of 8 bit ones (per the errata - which went unacknowledged by Microchip for years.)

**Reply by farfarfromhome** [reply]
Ok I have verified it does work as you suggested but with period 1 (two clocks) and duty cycle 1 (one clock).  This is consistent with spec since  output goes high at bottom and clears on compare.   My pin is saved and no (at)tiny wires!

Thanks again for all your help.
